### PR TITLE
[WIP] refactored communication creation in MOE

### PR DESCRIPTION
**Pull request by 1SAA**
- refactored communication creation in MOE to support PR-MOE
- changed moe gradient handler to adapt new feature

### CODE REVIEW

**Comment by FrankLeeeee** [comment]
Hi @1SAA , can you avoid having a large number of files in one PR next time? It is tough to review in detail with too many changes.

**Comment by 1SAA** [comment]
I understand. But this feature needs to change the fundamental communication functions when using MoE. 
Any code about MoE can't run properly or even crashed, when any fundamental changes occur.
In order to make MoE example runs properly, I have to change so much code. I'm sorry about that.

File: colossalai/engine/gradient_handler/_moe_gradient_handler.py [hunk 1]
```diff
@@ -21,41 +42,31 @@ def handle_gradient(self):
...
-
-                dist.all_reduce(
-                    coalesced, group=gpc.get_group(ParallelMode.DATA))
-                for buf, synced in zip(grads, _unflatten_dense_tensors(
-                        coalesced, grads)):
-                    buf.copy_(synced)
+            dist_dict = moe_env.information
+            diff_params = dict()
+
+            diff_params[1] = []
```
Comment (feifeibear) [inline colossalai/engine/gradient_handler/_moe_gradient_handler.py:51]: no one can understand these code.

File: colossalai/context/moe_context.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                self.ep_group = group
+
+        for j in range(ep_size):
+            ranks = [i * ep_size + j for i in range(dp_size)]
+            group = dist.new_group(ranks)
+            if rank in ranks:
+                self.dp_group = group
+
+
+class MoeContext:
```
[the remark this answers is not in the document]
Comment (1SAA) [reply colossalai/context/moe_context.py:48]: I tried to derive it from Singleton in colossalai.utils. But some weird reference problem has raised.

File: colossalai/context/moe_context.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+class MoeInfo:
+    """Moe parallelism information, storing parallel sizes and groups.
+    """
+
+    def __init__(self, ep_size: int, dp_size: int):
+        self.ep_size = ep_size
+        self.dp_size = dp_size
+        self.ep_group = None
+        self.dp_group = None
+
```
Comment (feifeibear) [inline colossalai/context/moe_context.py:15]: add assert to your consumptions! make sure the program can run in the env out of your expectation.
assert gpc.get_group(ParallelMode.TENSOR) == 1, "You error info"

File: colossalai/context/moe_context.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        if ep_size == 1:
+            from colossalai.core import global_context as gpc
+            self.ep_group = gpc.get_group(ParallelMode.TENSOR)
+            self.dp_group = gpc.get_group(ParallelMode.DATA)
+            return
+
+        if dp_size == 1:
+            from colossalai.core import global_context as gpc
+            self.ep_group = gpc.get_group(ParallelMode.DATA)
+            self.dp_group = gpc.get_group(ParallelMode.TENSOR)
```
Comment (feifeibear) [inline colossalai/context/moe_context.py:29]: so wired....
self.dp_group is diff from gpc's DP group
Add comments to explain what is self.dp_group

File: colossalai/context/moe_context.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    def get_instance():
+        if MoeContext.__instance is None:
+            MoeContext.__instance = MoeContext()
+        return MoeContext.__instance
+
+    def __init__(self):
+        self.world_size = 1
+        # Users may want to set maximum expert parallel size smaller than the world size
+        # since very low bandwidth across nodes may constrain the performance of MoE
+        # When we have a maximum expert parallel size, we have a minimum data parallel size naturally
```
Comment (feifeibear) [inline colossalai/context/moe_context.py:64]: what do you want to say? any suggestions to the max_ep_size setting?

File: colossalai/context/moe_context.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    def information(self):
+        return self._info_dict
+
+    @property
+    def is_initialized(self):
+        return self.has_setup
+
+    def setup(self, seed: int, enable_cuda: bool = True):
+        from colossalai.core import global_context as gpc
+
```
Comment (feifeibear) [inline colossalai/context/moe_context.py:83]: make sure we do not setup twice!!!

File: colossalai/context/moe_context.py (hunk 5 above)
Comment (1SAA) [reply colossalai/context/moe_context.py:64]: Just tell the function of max_ep_size

**Pull request closed without merging**